src/conn/pool/mod.rs

```diff
@@ -192,12 +192,32 @@ struct Inner {
     extra_wakeups: atomic::AtomicUsize,
 
     // only used to spawn the recycler the first time we're in async context
-    maker: Mutex<Option<mpsc::UnboundedReceiver<Option<Conn>>>>,
+    maker: Mutex<Option<(mpsc::UnboundedReceiver<Option<Conn>>, usize)>>,
 }
 
 impl Inner {
+    fn spawn_recycler_if_needed(self: Arc<Self>) {
+        let mut lock = self.maker.lock().unwrap();
+        if let Some((dropped, min)) = lock.take() {
+            // we're the first connection!
+            tokio::spawn(Recycler {
+                inner: self.clone(),
+                discard: FuturesUnordered::new(),
+                discarded: 0,
+                cleaning: FuturesUnordered::new(),
+                dropped,
+                min,
+                eof: false,
+            });
+        }
+    }
+
     fn wake(&self, mut readied: usize) {
         if readied == 0 {
+            while let Ok(task) = self.wake.pop() {
```

Comment:
Hmm, this doesn't quite seem right to me. If `readied == 0`, then no tasks should be woken up, because there are no connections available. If this is really necessary, that suggests to me that we need to have this `while` when we are polling the disconnect future.

Reply:
@jonhoo please look into the new version.